fortuna/hallucination/base.py

```diff
@@ -0,0 +1,335 @@
+        )
+
+    def _compute_scores_embeddings_which_choices(
+        self,
+        texts: Union[List[str], List[List[str]]],
+        contexts: List[str],
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+        scores = []
+        embeddings = []
+        which_choices = []
+
+        for text, context in tqdm(zip(texts, contexts)):
+            context_inputs = self.tokenizer(context, return_tensors="pt").to(
+                self.generative_model.device
+            )
+            len_context_inputs = len(context_inputs)
+            if isinstance(text, list):
+                _scores = []
+                _embeddings = []
+
+                for _text in text:
+                    _text_inputs = self.tokenizer(_text, return_tensors="pt").to(
+                        self.generative_model.device
+                    )
+                    _inputs = {
+                        k: torch.cat((context_inputs[k], v), dim=1)
+                        for k, v in _text_inputs.items()
+                    }
+
+                    with torch.no_grad():
+                        __logits = self.generative_model(
+                            input_ids=_inputs["input_ids"],
+                            attention_mask=_inputs["attention_mask"],
+                        ).logits
+
+                    _scores.append(
+                        self.scoring_fn(
+                            logits=__logits,
+                            labels=_inputs["input_ids"],
+                            init_pos=len_context_inputs,
+                        )
+                        .cpu()
+                        .numpy()
+                    )
+                    _embeddings.append(__logits.mean(1).cpu().numpy())
+
+                which_choice = np.argmax(_scores)
+                which_choices.append(which_choice)
+                scores.append(_scores[which_choice])
+                embeddings.append(_embeddings[which_choice])
+
+            elif isinstance(text, str):
+                text_inputs = self.tokenizer(text, return_tensors="pt").to(
+                    self.generative_model.device
+                )
+                inputs = {
+                    k: torch.cat((context_inputs[k], v), dim=1)
+                    for k, v in text_inputs.items()
+                }
+
+                with torch.no_grad():
+                    _logits = self.generative_model(
+                        input_ids=inputs["input_ids"],
+                        attention_mask=inputs["attention_mask"],
+                    ).logits
+                    embeddings.append(_logits.mean(1).cpu().numpy())
+
+                scores.append(
+                    self.scoring_fn(
+                        logits=_logits,
+                        labels=inputs["input_ids"],
+                        init_pos=len_context_inputs,
+                    )
+                    .cpu()
+                    .numpy()
+                )
+
+            else:
+                raise ValueError(
```

Comment:
after refactoring, you might ignore this case and let the user run into the exception. you can't protect against all wrong inputs in this way